New muting/appeal channels system

le-potate opened the pull request:
New muting system. Users get muted either by using the mute command or by being given the muted role. Once this happens, all other roles are saved and removed, and an appeal channel for that user is created. When unmuted (either using unmute command or removing roles), the previous roles can be given back, and the appeal channel is deleted. I've tried to make it so that this doesn't break completely if the channel is deleted manually, the role is removed while the bot isn't on, etc.

The part that saves roles when a user is muted was mostly written by @davidlougheed

## How Has This Been Tested?
Light testing to ensure the main functions work and this doesn't break the bot completely. We should go ahead and merge this ASAP as this is needed quickly, but we should keep in mind that we may need to work on this again in the near future.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.

le-potate committed:
Mute/unmute command

Command that mutes users, removing and saving all their roles and creating an appeal channel for them.
The part that saves roles when muting was from @davidlougheed, he will lose credit for most of it since it was moved to a different file and had never made it to master
Since there is some overlap with role restoration etc, some parts from roles.py (written by david) were moved to a utils file
All references to penalty are now renamed muted

le-potate committed:
lint

le-potate committed:
update poetry.lock

not sure if this will work as I'm having some issues

davidlougheed commented on cogs/utils/paginator.py:
```diff
@@ -93,11 +97,20 @@ def __init__(
+            self.bot = bot
+            self.guild = guild
+            self.channel = channel
+            self.user = user
```
personally, I'd rewrite this to allow ctx to either be the discordpy context or a custom context class which mimics these properties, which can be passed in as an alternative
so ctx would have the signature of roughly
```py
from typing import Union
# ...
ctx: Union[discord.Context, MockContext]
```

le-potate replied:
resolved in 8af4976

le-potate committed:
Remove print statement

davidlougheed commented on cogs/mod.py:
```diff
@@ -111,6 +141,222 @@ async def finalize_crabbo(self, ctx):
+                log_message = f"User {message.author} deleted message {message.id} in the appeal channel for muted user {muted_user.mention} ({muted_user}), {message.channel.mention}"
+            await self.appeals_log_channel.send(log_message)
+
+    # the next two functions are used to save the roles
```
i only see one function

le-potate replied:
resolved in 8af4976

davidlougheed commented on Martlet.schema:
```diff
@@ -78,11 +78,13 @@ CREATE TABLE IF NOT EXISTS `PreviousRoles` (
+-- and stores old roles akin to the way PreviousRoles does it
+CREATE TABLE IF NOT EXISTS `MutedUsers` (
+    `UserID`              INTEGER UNIQUE,
+	`AppealChannelID`     INTEGER UNIQUE,
```
whats up with the indents here

le-potate replied:
resolved in 8af4976

davidlougheed commented on cogs/roles.py:
```diff
@@ -248,246 +246,40 @@ async def create_role(self, ctx, *, role: Optional[str] = None):
+            save_existing_roles(self, user, muted=True)
+        else:
+            # Save existing roles
+            save_existing_roles(self, user)
```
can this maybe be factored into
```py
@commands.Cog.listener()
async def on_member_remove(self, user: discord.Member):
    save_existing_roles(self, user, muted=is_in_muted_table(self, user))
```

also passing self as an explicit param is sort of non-conventional; you could make a mixin class and use multiple inheritance, or pass in the DB path or the bot object instead perhaps

le-potate replied:
resolved in 8af4976

le-potate committed:
requested changes and small fixes

notably, this adds a mock context utility that can be used to replace discord context in context-less situation

davidlougheed committed:
fix some minor issues and improve code style/type hinting

davidlougheed committed:
improve role restoration type hints

davidlougheed committed:
lint

le-potate committed:
Merge branch 'dev' into appeals
